folding-schemes/src/frontend/noname/mod.rs

```diff
@@ -0,0 +1,228 @@
+use crate::Error;
+use ark_noname::sonobe::NoNameSonobeCircuit;
+use ark_r1cs_std::alloc::AllocVar;
+use ark_r1cs_std::fields::fp::FpVar;
+use ark_relations::r1cs::{ConstraintSynthesizer, SynthesisError};
+use num_bigint::BigUint;
+use std::marker::PhantomData;
+
+use super::FCircuit;
+use ark_ff::PrimeField;
+use ark_noname::utils::compile_source_code;
+use noname::backends::{r1cs::R1CS as R1CSNoName, BackendField};
+use noname::witness::CompiledCircuit;
+
+pub mod utils;
+
+use utils::*;
+
+#[derive(Debug, Clone)]
+pub struct NoNameFCircuit<F: PrimeField, BF: BackendField> {
+    pub state_len: usize,
+    pub external_inputs_len: usize,
+    pub circuit: CompiledCircuit<R1CSNoName<BF>>,
+    _f: PhantomData<F>,
+}
+
+impl<F: PrimeField, BF: BackendField> FCircuit<F> for NoNameFCircuit<F, BF> {
+    type Params = (String, usize, usize);
+
+    fn new(params: Self::Params) -> Result<Self, crate::Error> {
+        let (code, state_len, external_inputs_len) = params;
+        let compiled_circuit = compile_source_code::<BF>(&code).map_err(|_| {
+            Error::Other("Encountered an error while compiling a noname circuit".to_owned())
+        })?;
+        Ok(NoNameFCircuit {
+            state_len,
+            external_inputs_len,
+            circuit: compiled_circuit,
+            _f: PhantomData,
+        })
+    }
+
+    fn state_len(&self) -> usize {
+        self.state_len
+    }
+
+    fn external_inputs_len(&self) -> usize {
+        self.external_inputs_len
+    }
+
+    fn step_native(
+        &self,
+        _i: usize,
+        z_i: Vec<F>,
+        external_inputs: Vec<F>,
+    ) -> Result<Vec<F>, crate::Error> {
+        // compute witness using noname witness generator and outputs the next step
+        let external_inputs_json = match !external_inputs.is_empty() {
+            false => None,
+            true => {
+                let json =
+                    serde_json::to_string(&NoNameJSONExternalInputs::from(external_inputs.clone()))
+                        .map_err(|e| Error::JSONSerdeError(e.to_string()))?;
+                Some(json)
+            }
+        };
```

Comment:
When you deal with binary option, you should use `if-else` instead of match.

Reply:
Indeed, thanks for suggesting. Made sure this was the case in the latest PR change.